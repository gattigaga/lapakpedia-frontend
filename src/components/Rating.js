import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Star from "react-icons/lib/io/star";

export const StyledStar = styled(Star)`
  color: ${props => (props.isFilled ? "#ff8800" : "#bbb")};
  ${props => !props.isDisabled && "cursor: pointer"};
`;

class Rating extends Component {
  constructor(props) {
    super(props);

    this.state = {
      hoveredValue: props.value
    };

    this.handleHover = this.handleHover.bind(this);
  }

  /**
   * Handle hovered star value
   *
   * @param {number} value - Star value
   * @memberof Rating
   */
  handleHover(value) {
    this.setState({ hoveredValue: value });
  }

  render() {
    const { hoveredValue } = this.state;
    const { value, onClick, isDisabled } = this.props;

    return (
      <div>
        {[...Array(5)].map((star, index) => {
          const newValue = index + 1;

          return (
            <StyledStar
              key={index}
              isFilled={newValue <= hoveredValue}
              isDisabled={isDisabled}
              onClick={() => onClick(newValue)}
              onMouseEnter={() => !isDisabled && this.handleHover(newValue)}
              onMouseLeave={() => !isDisabled && this.handleHover(value)}
            />
          );
        })}
      </div>
    );
  }
}

Rating.propTypes = {
  value: PropTypes.number,
  onClick: PropTypes.func,
  isDisabled: PropTypes.bool
};

export default Rating;
