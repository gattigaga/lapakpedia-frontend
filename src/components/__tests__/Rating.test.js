import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Rating, { StyledStar } from "components/Rating";

describe("Rating", () => {
  const setup = propOverrides => {
    const props = {
      onClick: jest.fn(),
      value: 3,
      ...propOverrides
    };

    const wrapper = shallow(<Rating {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders in disable state", () => {
    const { wrapper } = setup({ isDisabled: true });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should calls 'onClick' callback while a star clicked", () => {
    const { props, wrapper } = setup();
    const value = 4;

    wrapper
      .find(StyledStar)
      .at(value - 1)
      .simulate("click");

    expect(props.onClick).toBeCalledWith(value);
  });
});
