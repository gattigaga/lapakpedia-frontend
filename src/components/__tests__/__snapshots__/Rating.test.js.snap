// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should renders default 1`] = `
<div>
  <Styled(IoStar)
    isFilled={true}
    key="0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isFilled={true}
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isFilled={true}
    key="2"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isFilled={false}
    key="3"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isFilled={false}
    key="4"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
</div>
`;

exports[`Rating should renders in disable state 1`] = `
<div>
  <Styled(IoStar)
    isDisabled={true}
    isFilled={true}
    key="0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isDisabled={true}
    isFilled={true}
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isDisabled={true}
    isFilled={true}
    key="2"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isDisabled={true}
    isFilled={false}
    key="3"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
  <Styled(IoStar)
    isDisabled={true}
    isFilled={false}
    key="4"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={14}
  />
</div>
`;

exports[`Rating should renders with custom size 1`] = `
<div>
  <Styled(IoStar)
    isFilled={true}
    key="0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={32}
  />
  <Styled(IoStar)
    isFilled={true}
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={32}
  />
  <Styled(IoStar)
    isFilled={true}
    key="2"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={32}
  />
  <Styled(IoStar)
    isFilled={false}
    key="3"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={32}
  />
  <Styled(IoStar)
    isFilled={false}
    key="4"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={32}
  />
</div>
`;
